agent/agent_test.go

```diff
@@ -3490,6 +3490,36 @@ func TestAgent_ReloadConfigTLSConfigFailure(t *testing.T) {
 	require.Len(t, tlsConf.RootCAs.Subjects(), 1)
 }
 
+func TestAgent_ReloadConfigUIConfig(t *testing.T) {
+	t.Parallel()
+	dataDir := testutil.TempDir(t, "agent") // we manage the data dir
+	hcl := `
+		data_dir = "` + dataDir + `"
+		ui_config {
+			enabled = true // note that this is _not_ reloadable
+			metrics_provider = "foo"
+		}
+	`
+	a := NewTestAgent(t, hcl)
+	defer a.Shutdown()
```

Comment:
This is a very expensive test for what is essentially testing that an atomic works. Do we really need a running agent to be able to test this?

Reply:
🤔 This is really a test of the plumbing and glue code around reloading. It follows the pattern of the other reload tests. IIRC for a long time we didn't have tests for much of this stuff and often had bugs where code had been added to a reload method but some part of the plumbing missed.

Sadly since the agent isn't more modular it's pretty hard to test any of the plumbing around one without starting one.

I'm with you on wanting to make tests faster and reduce the number of times we spin up a full agent but that seems like something we need to tackle by actually breaking up the agent rather than just not testing new stuff we do add to it?

> Do we really need a running agent to be able to test this?

Maybe not but if we hacked together an agent struct bypassing all the normal startup code and called an internal method to reload a pointer... then it would be only a test of whether the atomic.Value works as expected! If we had a way to start an agent using the normal code paths but skipping some expensive elements like not starting serf or raft maybe that might work?

Overall though, as discussed above I'd rather that this config reload plumbing was all encapsulated in a separate package and not on the agent itself but that's out of scope here.

Reply:
😄  - this comment made me realise that this test was actually failing...

Because I'd forgotten to actually add the `Store` in the `reloadConfigInternal` method!

Perhaps there is a cheaper way if we change the reload logic a little like you propose in #8623 since then reloaders are not relying on any other state in the agent to be a certain way and we can have a cheaper test that just ensure s all the expected reloaders have indeed been registered after a "normal" startup?

Plumbing tests always seem wasteful since the code is trivial unless it's not there!